### PR TITLE
Use python scripts for building cpp, conda package and docs

### PR DESCRIPTION
**Pull request by nvaytet**
**Changes:**
- Replace `tools/make_and_install.*` scripts with `tools/build_cpp.py`
- Replace `conda/build.sh` and `conda/bld.bat` with `tools/build_conda.py` and a `script` line in `conda/meta.yml`
- Replace `cmake/docs_build.*` with `docs/make_docs.py`

I put `build_conda.py` and `build_cpp.py` in the same folder so that `build_conda.py` could import from `build_cpp.py`.

**Benefits:**
- use `os.path.join` to take care of forward or backward slashes in file paths
- use `shutil.move` and `glob.glob` to take care of moving files and folders in a consistent manner
- use `os.environ.get` to get environment variables in a platform-independent manner
- use `os.makedirs` to create folder in a platform-independent manner

**Use the tools locally:**
- Use `python tools/build_cpp.py` to build and run the C++ tests
- Use `python docs/make_docs.py` to build the documentation

See companion PRs in `scippneutron` (https://github.com/scipp/scippneutron/pull/80) and `pipelines` (https://github.com/scipp/pipelines/pull/4)

**Note that the PR in `pipelines` must be merged first.**

### CODE REVIEW

**Comment by nvaytet** [comment]
This is now ready although I am still not happy with all the duplication in `build_cpp.py`, `build_conda.py` (e.g. the `FileMover`) and `make_docs.py` between `scipp`, `scippneutron` and `pipelines`.

I however don't really know what would be the best way to remove the duplication would be.
- Using a git submodule sounds a little messy trying to get all the paths rights?
- Putting the common code into a small conda package that would then be part of the `scipp-developer.yml` env file would solve the paths issue but might make it more difficult to make changes to the pipelines?

Although on the second item, we could deploy some conda packages under a e.g. `test` label on the scipp channel even for pull requests, so that we can test things out?

File: tools/build_cpp.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                shell=shell)
+
+    # Compile C++ tests
+    run_command(['cmake', '--build', '.', '--target', 'all-tests'] +
+                build_flags,
+                shell=shell)
+
+    # Compile Python library
+    run_command(['cmake', '--build', '.', '--target', 'install'] + build_flags,
+                shell=shell)
```
Comment (SimonHeybrock) [inline tools/build_cpp.py:110]:
```suggestion
    for target in ['all-benchmarks', 'all-tests', 'install']:
        run_command(['cmake', '--build', '.', '--target', target] + build_flags, shell=shell)
```
and similar below for the test runs?

**Review by SimonHeybrock — COMMENTED** [review]
Nice work, changes make sense.

**Comment by OwenArnold** [comment]
> This is now ready although I am still not happy with all the duplication in `build_cpp.py`, `build_conda.py` (e.g. the `FileMover`) and `make_docs.py` between `scipp`, `scippneutron` and `pipelines`.
> 
> I however don't really know what would be the best way to remove the duplication would be.
> 
> * Using a git submodule sounds a little messy trying to get all the paths rights?
> * Putting the common code into a small conda package that would then be part of the `scipp-developer.yml` env file would solve the paths issue but might make it more difficult to make changes to the pipelines?
> 
> Although on the second item, we could deploy some conda packages under a e.g. `test` label on the scipp channel even for pull requests, so that we can test things out?

I would vote for the second option.

**Pull request merged**